### ISSUE TITLE
Reuse of model name from package breaks schema generation

### ISSUE THREAD

**Issue by jackmpcollins** [pydantic/pydantic#10105]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

The `openai` Python package contains the definitions

```python
class FunctionCall(TypedDict, total=False):
    arguments: Required[str]
    name: Required[str]


class ChatCompletionAssistantMessageParam(TypedDict, total=False):
    # ...
    function_call: Optional[FunctionCall]
```

In my code I also have a `FunctionCall` class. When I try to create a new pydantic model containing a field that has type `ChatCompletionAssistantMessageParam`, it results in a `PydanticSchemaGenerationError` exception that indicates pydantic is trying to use my code's `FunctionCall` instead of the openai package `FunctionCall`.

```python
from openai.types.chat import ChatCompletionAssistantMessageParam
from pydantic import BaseModel


class FunctionCall:
    ...


class Test(BaseModel):
    test: ChatCompletionAssistantMessageParam


# PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class '__main__.FunctionCall'>.
#  Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
```

It seems like the namespaces are not correctly being kept separate. I would expect my reuse of class names to have no effect on already-defined pydantic models.

---

I can't seem to recreate the same issue without importing from another package. For example this code does not reproduce it.

```python
from typing_extensions import TypedDict
from pydantic import BaseModel


class Test(TypedDict, total=False):
    value: int


class TestContainer(BaseModel):
    test: Test


class Test:
    ...


class TestContainerContainer(BaseModel):
    test_container: TestContainer
```

### Example Code

```Python
from openai.types.chat import ChatCompletionAssistantMessageParam
from pydantic import BaseModel


class FunctionCall:
    ...


class Test(BaseModel):
    test: ChatCompletionAssistantMessageParam
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=true
                 install path: /Users/jack/Code/notebooks/.venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Jul 15 2023, 09:54:16) [Clang 14.0.3 (clang-1403.0.22.14.1)]
                     platform: macOS-14.6.1-arm64-arm-64bit
             related packages: pydantic-settings-2.3.4 typing_extensions-4.12.2
                       commit: unknown
```

**Comment by Viicos** [comment]
I've tried reproducing the issue in multiple ways without any success. Could you please provide the output of the following code:

```python
from openai.types.chat import ChatCompletionAssistantMessageParam

print(ChatCompletionAssistantMessageParam.__annotations__['function_call'])
print(ChatCompletionAssistantMessageParam.__annotations__['function_call'].__forward_module__)

from pydantic import BaseModel


class FunctionCall:
    ...

import sys

print(sys.modules['openai.types.chat.chat_completion_assistant_message_param'].__dict__['FunctionCall'])


class Test(BaseModel):
    test: ChatCompletionAssistantMessageParam
```

**Comment by jackmpcollins** [comment]
I created a new virtualenv, installed the latest openai and pydantic, and can reproduce. The output is

```
ForwardRef('Optional[FunctionCall]', module='openai.types.chat.chat_completion_assistant_message_param')
openai.types.chat.chat_completion_assistant_message_param
<class 'openai.types.chat.chat_completion_assistant_message_param.FunctionCall'>
```

Package versions
```
% rye list
annotated-types==0.7.0
anyio==4.4.0
certifi==2024.7.4
distro==1.9.0
h11==0.14.0
httpcore==1.0.5
httpx==0.27.0
idna==3.7
jiter==0.5.0
openai==1.40.6
pydantic==2.8.2
pydantic-core==2.20.1
sniffio==1.3.1
-e file:///Users/jack/Code/test-pydantic-bug
tqdm==4.66.5
typing-extensions==4.12.2
```

Pydantic version details
```
% python -c "import pydantic.version; print(pydantic.version.version_info())"

             pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=true
                 install path: /Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic
               python version: 3.12.4 (main, Jul 25 2024, 22:11:22) [Clang 18.1.8 ]
                     platform: macOS-14.6.1-arm64-arm-64bit
             related packages: typing_extensions-4.12.2
                       commit: unknown
```

<details>
<summary>Full output</summary>

```
% rye run python
Python 3.12.4 (main, Jul 25 2024, 22:11:22) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from openai.types.chat import ChatCompletionAssistantMessageParam
>>>
>>> print(ChatCompletionAssistantMessageParam.__annotations__['function_call'])
ForwardRef('Optional[FunctionCall]', module='openai.types.chat.chat_completion_assistant_message_param')
>>> print(ChatCompletionAssistantMessageParam.__annotations__['function_call'].__forward_module__)
openai.types.chat.chat_completion_assistant_message_param
>>>
>>> from pydantic import BaseModel
>>>
>>>
>>> class FunctionCall:
...     ...
...
>>> import sys
>>>
>>> print(sys.modules['openai.types.chat.chat_completion_assistant_message_param'].__dict__['FunctionCall'])
<class 'openai.types.chat.chat_completion_assistant_message_param.FunctionCall'>
>>>
>>>
>>> class Test(BaseModel):
...     test: ChatCompletionAssistantMessageParam
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 205, in __new__
    complete_model_class(
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 534, in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/main.py", line 643, in __get_pydantic_core_schema__
    return handler(source)
           ^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 512, in generate_schema
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 784, in _generate_schema_inner
    return self._model_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 591, in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 947, in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1134, in _common_field_schema
    schema = self._apply_annotations(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1890, in _apply_annotations
    schema = get_inner_schema(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1871, in inner_handler
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 789, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 837, in match_type
    return self._typed_dict_schema(obj, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1336, in _typed_dict_schema
    fields[field_name] = self._generate_td_field_schema(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 930, in _generate_td_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1134, in _common_field_schema
    schema = self._apply_annotations(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1890, in _apply_annotations
    schema = get_inner_schema(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1871, in inner_handler
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 789, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 871, in match_type
    return self._match_generic_type(obj, origin)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 895, in _match_generic_type
    return self._union_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1207, in _union_schema
    choices.append(self.generate_schema(arg))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 512, in generate_schema
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 789, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 875, in match_type
    return self._unknown_type_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jack/Code/test-pydantic-bug/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 415, in _unknown_type_schema
    raise PydanticSchemaGenerationError(
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class '__main__.FunctionCall'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.8/u/schema-for-unknown-type
```

</details>

**Comment by Viicos** [comment]
Turns out this was fixed by https://github.com/pydantic/pydantic/pull/10063. It will be included in the next 2.9 release.
